azurelinuxagent/common/protocol/imds.py

```diff
@@ -1,22 +1,155 @@
+            
+            "16.04-LTS": { "Match": ".*" },
+            "16.04.0-LTS": { "Match": ".*" },
+            
+            "18.04-LTS": { "Match": ".*" },
+            "20.04-LTS": { "Match": ".*" },
+            "22.04-LTS": { "Match": ".*" }
+        }
+    },
+    "COREOS": {
+        "COREOS": {
+            "STABLE": { "Minimum": "494.4.0" }
+        }
+    },
+    "CREDATIV": {
+        "DEBIAN": { "Minimum": "7" }
+    },
+    "OPENLOGIC": {
+        "CENTOS": {
+            "Minimum": "6.3",
+            "7-LVM": { "Match": ".*" },
+            "7-RAW": { "Match": ".*" }
+        },
+        "CENTOS-HPC": { "Minimum": "6.3" }
+    },
+    "REDHAT": {
+        "RHEL": { 
+            "Minimum": "6.7",
+            "7-LVM": { "Match": ".*" },
+            "7-RAW": { "Match": ".*" }
+        },
+        "RHEL-HANA": { "Minimum": "6.7" },
+        "RHEL-SAP": { "Minimum": "6.7" },
+        "RHEL-SAP-APPS": { "Minimum": "6.7" },
+        "RHEL-SAP-HANA": { "Minimum": "6.7" }
+    },
+    "SUSE": {
+        "SLES": {
+            "11-SP4": { "Match": ".*" },
+            "11-SP5": { "Match": ".*" },
+            "11-SP6": { "Match": ".*" },
+            "12-SP1": { "Match": ".*" },
+            "12-SP2": { "Match": ".*" },
+            "12-SP3": { "Match": ".*" },
+            "12-SP4": { "Match": ".*" },
+            "12-SP5": { "Match": ".*" },
+            "12-SP6": { "Match": ".*" }
+        },
+        "SLES-BYOS": {
+            "11-SP4": { "Match": ".*" },
+            "11-SP5": { "Match": ".*" },
+            "11-SP6": { "Match": ".*" },
+            "12-SP1": { "Match": ".*" },
+            "12-SP2": { "Match": ".*" },
+            "12-SP3": { "Match": ".*" },
+            "12-SP4": { "Match": ".*" },
+            "12-SP5": { "Match": ".*" },
+            "12-SP6": { "Match": ".*" }
+        },
+        "SLES-SAP": {
+            "11-SP4": { "Match": ".*" },
+            "11-SP5": { "Match": ".*" },
+            "11-SP6": { "Match": ".*" },
+            "12-SP1": { "Match": ".*" },
+            "12-SP2": { "Match": ".*" },
+            "12-SP3": { "Match": ".*" },
+            "12-SP4": { "Match": ".*" },
+            "12-SP5": { "Match": ".*" },
+            "12-SP6": { "Match": ".*" }
+        }
+    }
+}"""
+
+
+class ImageInfoMatcher(object):
+    def __init__(self, doc):
+        self.doc = json.loads(doc)
+
+    def is_match(self, publisher, offer, sku, version):
+        def _is_match_walk(doci, keys):
+            key = keys.pop(0).upper()
+            if key is None:
+                return False
+
+            if key not in doci:
+                return False
+
+            if 'Match' in doci[key] and re.match(doci[key]['Match'], keys[0]):
+                return True
+
+            if 'Minimum' in doci[key]:
+                try:
+                    return FlexibleVersion(keys[0]) >= FlexibleVersion(doci[key]['Minimum'])
+                except ValueError:
+                    pass
+
```

Comment:
Consider adding a "list" entry, whose value is a JSON list of strings; if keys[0] is in that list, return True. This truncates the search early in many cases, but lets you eliminate nearly all of your "Match": ".*" cases.

```
   "CANONICAL": {
        "UBUNTUSERVER": { 
            "List": [ "14.04.0-LTS", "14.04.1-LTS", "14.04.2-LTS", "14.04.3-LTS", etc ]
         }
    },
```

And...

```
    if 'List' in doci[key] and keys[0] in doci[key]['List']:
        return True
```